content/blog/2022-05-26-typed-assembly-language/index.md

```diff
@@ -0,0 +1,81 @@
+	- [x] AST
+		- [x] lexing
+		- [x] parsing
+		- [x] type checking
+	- [ ] IR
+		- [x] cps (continuation passing style) translation
+		- [ ] closure conversion
+		- [ ] abstract assembly
+	- [ ] Polymorphic types
+- [ ] typed assembly
+	- [ ] syntax specification
+	- [ ] type checking
+	
+	
+The entire project is written in Rust.
+
+- For the lexing stage, I used [logos](https://docs.rs/logos/latest/logos/), a lexer generator. The tokens are represented as the enum type `Token` in `lex.rs`. The tokens include all the SML keywords, integer and boolean literals, special characters like `(`, `[`, `:`, etc. The tokens also include binary operators `+`, `-`, `*`, as well as built-in functions `not`, `andalso`, `orelse`. Finally, it also supports built-in types `int`, `bool`, `unit` as well as polymorphic type variables.
+
+- For parsing, I wrote a grammar for the subset of sml to work with the parser generator [lalrpop](https://github.com/lalrpop/lalrpop). Since SML is completely specified, I was able to figure out the precedence hierarchy of expressions with not much difficulty. This took a considerable amount of time. As usual, the program is represented as an AST, with a mixture of Rust structs and enums. Since the expression types need to refer to itself, some of the struct fields needed to be boxed, like `Box<Expr>`, for example. Finally, each expression variant has a field `typ`, which is initialized to be `Typ::Unknown` in the parser will be filled in by the type checker as the type of the expression. 
+
+- AST type checking is done in the traditional way with a typing context, mapping identifiers to their types. Of course, the context needs to accomodate different "levels" in the AST. Since type checking might be needed for multiple stages in the compilation, I abstracted away a generic context structure.
+
+- The struct `Context<T: Clone>` lives in `util.rs`. It has functions `bind`, `get`, `enter`, and `exeuent`. Think of it as having "levels", corresponding to the parent-child relationship in any tree structure. `bind` binds an identifier to some `T` in the current level. `get` gets the value bound to some identifier. `enter` "saves" the current level and goes down one, and `exeuent` discards the current level and restores the previous level. Each level can see the mappings from all previous levels. For example, to check `e1 + e2` one would wrap the type checking of each of `e1` and `e2` between an `enter` and `exeuent`, and see if both expressions type-checked to `int`.
+
+
+- CPS translation follows pretty closely to the Morrisett paper. Instead of just expressions we now have `Expr` and `Value`. `Value` are like expressions but they cannot be further beta-reduced. The types are almost the same, except `Arrow` type is replaced by `Cont`, which doesn't have return types: CPS replaces all returns with a call to a continuation, so every function "returns" void.
```

Comment:
Can you expand a little more on what "pretty closely to the Morrisett paper" looks like? Overall, I think it would be great to assume that your audience is knowledgable about compilers and type systems but has not read the TAL papers (beyond a cursory understanding). Up to this point in the post, this audience has been pretty much satisfied; there's not too much TAL-specific. But it would be good to know exactly what this CPS translation looks like and if/how it differs from a "normal" CPSification.

Reply:
This part was meant more like giving credit to the paper, as my implementation translates directly from their CPS translation. I modified the writeup to make this clear.